api/v2/routes_ipns.go

```diff
@@ -153,6 +153,20 @@ func (api *API) pinIPNSHash(c *gin.Context) {
 	// IPFS Cluster doesn't, so to keep consistency with the rest of our endpoints
 	// we should only operate on a bare cidC
 	hash := strings.Split(hashToPin, "/")[len(strings.Split(hashToPin, "/"))-1]
+	upload, err := api.upm.FindUploadByHashAndUserAndNetwork(username, hash, "public")
+	// by this conditional if statement passing, it means the user has
+	// upload content matching this hash before, and we don't want to charge them
+	// so we should gracefully abort further processing
+	if err == nil || upload != nil {
+		Respond(
+			c,
+			http.StatusOK,
+			gin.H{
```

Comment:
this is really spread out 😋 I would try to compress it a bit for readability:

```go
Response(c,
   http.StatusOK,
   gin.H{
      // ...
   })
```

Reply:
yes that is a good point.